### DISCUSSION TITLE
Differential/Delta Temperature ?

### DISCUSSION THREAD

**Discussion by Ex3c4Def** [tjko/fanpico#43] [Q&A]
Hello, is there a way to control a fan speed based on a delta temperature, like for a watercooling setup :
WaterTemperature - AmbientTemperature ?

If not, can this be implemented?

**Comment by Ex3c4Def** [comment]
From a first look at the sources it seems nothing like that is implemented yet.

A way of implementing it could be adding virtual sensors to the existing sensor, but that seems to need changes at a lot of places.

A better way may be adding a new fan source. This must be configured in some way and could be evaluated e.g. in the calculate_pwm_duty() function.

**Reply by tjko** [reply]
Yes, there is nothing like this yet implemented.

Adding configurable "virtual" sensors sounds like a good idea.   A virtual sensor could support reporting temperature delta (from two sensors). 

It could also support other "functions" like average, max, and min.   Since some cases there may be couple sensors measuring temperature from different locations and often (for cooling) one might want to be tracking "max" temperature or average temperature...

**Reply by Ex3c4Def** [reply]
That sounds nice. 
But I have at the moment not deep enough looked into the firmware to have a good understanding how everything works together, and where changes for this are needed.
Sadly in the moment also not much free time

**Reply by tjko** [reply]
I was already looking into adding support for (software/driver) provide virtual sensor readings from the host (CPU temp, etc). I'll see if I can implement this at same time as well.

**Reply by Ex3c4Def** [reply]
With that i could also use one on my computer.

At first i want to use it more or less standalone on a 3d printer. 

Do you already any time estimation when this might get implemented?

Just ordered some boards on ebay ;)

**Reply by tjko** [reply]
Thanks for supporting FanPico project.

This is on top of my "to do" list for FanPico currently, it should be in next firmware release that I hope to have ready in next couple weeks.

**Comment by tjko** [comment]
Initial support for virtual sensors is now implemented. This adds 8 virtual sensors that can get their temperature from the host or alternatively these can be configured to be 'calculated' based on 2 or more of the _sensors_.   There is support for: "delta", "min", "max", and "average" functions in this mode.

Changes are in separate development tree: https://github.com/tjko/fanpico/tree/vsensor

I'd expect to merge these to main branch soon, after I've had time to do some testing that everything seems to work as expected.

**Reply by Ex3c4Def** [reply]
Nice. Boards have not arrived yet. So  can not test it already.
